File: Java/Degree-of-an-Array.java
/*
*
*
*
*/

class Solution {
    public int findShortestSubArray(int[] nums) {
        Map<Integer, Integer> left = new HashMap(), right = new HashMap(), count = new HashMap();
 
        //find the leftmost and rightmost index of each characters
        for(int i= 0;i<nums.length; i++)
        {
            int x =nums[i];
            if(left.get(x) == null) 
                left.put(x, i);
            
            right.put(x, i);
            count.put(x, count.getOrDefault(x, 0) + 1);
        }
        
        int ans = nums.length;
        int degree  = Collections.max(count.values());
        
        //find min distance between subarrays with same degree
        for(int x: count.keySet())
        {
            if(count.get(x) == degree)
            {
                ans = Math.min(ans , right.get(x) - left.get(x) + 1);
            }
        }
        return ans;   
    }
}

File: Python/degree_of_an_array.py
"""
Question: 
Given a non-empty array of non-negative integers nums,
the degree of this array is defined as the maximum 
frequency of any one of its elements.

Your task is to find the smallest possible length of a
(contiguous) subarray of nums, that has the same degree as nums.

Solution:
Iterate through the list of numbers whike keeping track of the
counter of leftmost and rightmost indices of the numbers
We then find the largest number and return the length  


Time Complexity:  O(n)
Space Complexity: O(n)
"""
import collections

class Solution(object):
    def findShortestSubArray(self, nums):
        """
        :type nums: list
        :rtype: int
        """
        counter = collections.Counter(nums)
        left_dict, right_dict = {}, {}
        for i, num in enumerate(nums):
            left_dict.setdefault(num, i)
            right_dict[num] = i
        degree = max(counter.values())
        result = min(right_dict[num]-left_dict[num]+1 \
                   for num in counter.keys() \
                   if counter[num] == degree)
        return result